zmq.hpp

```diff
@@ -615,8 +620,340 @@ inline void swap(context_t &a, context_t &b) ZMQ_NOTHROW {
+{
+    return mutable_buffer(mb.data(), (std::min)(mb.size(), n));
+}
+inline const_buffer buffer(const const_buffer& cb) noexcept
+{
+    return cb;
+}
+inline const_buffer buffer(const const_buffer& cb, size_t n) noexcept
+{
+    return const_buffer(cb.data(), (std::min)(cb.size(), n));
+}
+
 namespace detail
 {
+// utility functions for containers with data and size
+// data is nullptr if the container is empty
+template<class T> mutable_buffer buffar_mut_ds(T &data) noexcept
+{
+    return mutable_buffer(data.size() != 0u ? data.data() : nullptr,
+                          data.size() * sizeof(*data.data()));
+}
+template<class T> mutable_buffer buffar_mut_ds(T &data, size_t n_bytes) noexcept
+{
+    return mutable_buffer(data.size() != 0u ? data.data() : nullptr,
+                          (std::min)(data.size() * sizeof(*data.data()), n_bytes));
+}
+template<class T> const_buffer buffar_const_ds(const T &data) noexcept
+{
+    return const_buffer(data.size() != 0u ? data.data() : nullptr,
+                        data.size() * sizeof(*data.data()));
+}
+template<class T> const_buffer buffar_const_ds(const T &data, size_t n_bytes) noexcept
+{
+    return const_buffer(data.size() != 0u ? data.data() : nullptr,
+                        (std::min)(data.size() * sizeof(*data.data()), n_bytes));
+}
+template<class T> struct is_pod_like
+{
+    // NOTE: The networking draft N4771 section 16.11 requires
+    // T in the buffer functions below to be
+    // trivially copyable OR standard layout.
+    // Here we decide to be conservative and require both.
+    static constexpr bool value =
+      std::is_trivially_copyable<T>::value && std::is_standard_layout<T>::value;
+};
+} // namespace detail
+
+// C array
+template<class T, size_t N> mutable_buffer buffer(T (&data)[N]) noexcept
+{
+    static_assert(detail::is_pod_like<T>::value, "T must be POD");
+    static_assert(N > 0, "N > 0");
+    return mutable_buffer(static_cast<T *>(data), N * sizeof(T));
+}
+template<class T, size_t N>
+mutable_buffer buffer(T (&data)[N], size_t n_bytes) noexcept
+{
+    static_assert(detail::is_pod_like<T>::value, "T must be POD");
+    static_assert(N > 0, "N > 0");
+    return mutable_buffer(static_cast<T *>(data), (std::min)(N * sizeof(T), n_bytes));
+}
+template<class T, size_t N> const_buffer buffer(const T (&data)[N]) noexcept
+{
+    static_assert(detail::is_pod_like<T>::value, "T must be POD");
+    static_assert(N > 0, "N > 0");
+    return const_buffer(static_cast<const T *>(data), N * sizeof(T));
+}
+template<class T, size_t N>
+const_buffer buffer(const T (&data)[N], size_t n_bytes) noexcept
+{
+    static_assert(detail::is_pod_like<T>::value, "T must be POD");
+    static_assert(N > 0, "N > 0");
+    return const_buffer(static_cast<const T *>(data),
+                        (std::min)(N * sizeof(T), n_bytes));
+}
+// std::array
+template<class T, size_t N> mutable_buffer buffer(std::array<T, N> &data) noexcept
```

Comment:
Why are these overloads not using the functions from detail?

Is it necessary to provide separate overloads for array, vector and basic_string specifically, all of which could be implemented in the same way, as they are ContiguousContainers (see comment above)? That would be much less redundant and less verbose.

Reply:
I'm just following the spec as defined. I think it is like this because these are the types that are known to be contiguous in the standard library.
We could simplify this a lot by implementing one function, maybe using 
1. SFINAE for data and size (might be risky if called with some external type that happens to not be contiguous, but that may be the users fault anyway), or 
2. checking if the type is one of the known standard library types to be contiguous (a bit more verbose, but safe)

There are always the edge cases of C arrays and string::data() pre C++17 is always const.

Reply:
You mean the Networking TS here? Ok, maybe leave it like this then. I am not sure, it might get complex to SFINAE this without C++20 concepts, and maybe that's why the Networking TS does it like this as well, and it has not yet been updated to use C++20 language features?

It was just an idea if it had reduced complexity.